### ISSUE TITLE
你好，因为代码兼容性问题，我将register_backward_hook修改成了register_full_forward_hook，并对v_tgr的返回进行了一些修改，可是最后执行结果有差异。  另外我在启用PatchOut之后攻击成功率反而降低了，感觉奇怪。这是我的v_tgr的返回，以及我的环境和执行结果： 

### ISSUE THREAD

**Issue by guox18** [jpzhang1810/TGR#5]
def v_tgr(module, grad_in, grad_out, gamma):
            mask = torch.ones_like(grad_in[0]) * gamma
            out_grad = mask * grad_in[0][:]

            if self.model_name in ['visformer_small']:
                B,C,H,W = grad_in[0].shape
                out_grad_cpu = out_grad.data.clone().cpu().numpy().reshape(B,C,H*W)
                max_all = np.argmax(out_grad_cpu[0,:,:], axis = 1)
                max_all_H = max_all//H
                max_all_W = max_all%H
                min_all = np.argmin(out_grad_cpu[0,:,:], axis = 1)
                min_all_H = min_all//H
                min_all_W = min_all%H
                out_grad[:,range(C),max_all_H,max_all_W] = 0.0
                out_grad[:,range(C),min_all_H,min_all_W] = 0.0

            if self.model_name in ['vit_base_patch16_224', 'pit_b_224', 'cait_s24_224']:
                c = grad_in[0].shape[2]
                out_grad_cpu = out_grad.data.clone().cpu().numpy()
                max_all = np.argmax(out_grad_cpu[0,:,:], axis = 0)
                min_all = np.argmin(out_grad_cpu[0,:,:], axis = 0)
                    
                out_grad[:,max_all,range(c)] = 0.0
                out_grad[:,min_all,range(c)] = 0.0
            return tuple([out_grad] + list(grad_in[1:]))

环境：
cuda11.7 rtx3090

Package                  Version
------------------------ ----------
advertorch               0.2.3
astor                    0.8.1
backcall                 0.2.0
cachetools               5.3.1
certifi                  2022.12.7
charset-normalizer       3.1.0
colorama                 0.4.6
cpplint                  1.6.1
cycler                   0.11.0
dataclasses              0.6
debugpy                  1.7.0
decorator                5.1.1
dill                     0.3.7
eagerpy                  0.30.0
entrypoints              0.4
filelock                 3.12.2
fonttools                4.38.0
foolbox                  2.0.0rc0
fsspec                   2023.1.0
future                   0.18.3
gast                     0.2.2
gitdb                    4.0.10
GitPython                3.1.31
glog                     0.3.1
google-auth              2.22.0
google-auth-oauthlib     0.4.6
google-pasta             0.2.0
graphviz                 0.20.1
grpcio                   1.56.2
h5py                     3.8.0
huggingface-hub          0.16.4
idna                     3.4
imageio                  2.31.2
importlib-metadata       6.7.0
ipykernel                6.16.2
ipython                  7.34.0
jedi                     0.19.0
joblib                   1.3.2
jupyter_client           7.4.9
jupyter_core             4.12.0
Keras-Applications       1.0.8
Keras-Preprocessing      1.1.2
kiwisolver               1.4.4
Markdown                 3.4.4
MarkupSafe               2.1.3
matplotlib               3.5.3
matplotlib-inline        0.1.6
nest-asyncio             1.5.8
netron                   7.1.6
networkx                 2.6.3
numpy                    1.21.6
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-nvrtc-cu11   11.7.99
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96
oauthlib                 3.2.2
opencv-python            4.8.0.74
opt-einsum               3.3.0
packaging                23.1
pandas                   1.3.5
parso                    0.8.3
pexpect                  4.8.0
pickleshare              0.7.5
Pillow                   9.5.0
pip                      23.2.1
prompt-toolkit           3.0.39
protobuf                 3.20.3
psutil                   5.9.5
ptyprocess               0.7.0
pyasn1                   0.5.0
pyasn1-modules           0.3.0
Pygments                 2.16.1
pyparsing                3.0.9
python-dateutil          2.8.2
python-gflags            3.1.2
pytz                     2023.3
PyWavelets               1.3.0
PyYAML                   6.0.1
pyzmq                    25.1.1
randomgen                1.23.1
requests                 2.30.0
requests-oauthlib        1.3.1
rsa                      4.9
safetensors              0.3.3
scikit-image             0.19.3
scikit-learn             1.0.2
scipy                    1.7.3
setuptools               65.6.3
six                      1.16.0
smmap                    5.0.0
tensorboard              1.15.0
tensorboard-data-server  0.6.1
tensorboard-plugin-wit   1.8.1
tensorboardX             2.6.2.2
tensorflow-estimator     1.15.1
tensorflow-gpu           1.15.0
termcolor                2.3.0
threadpoolctl            3.1.0
tifffile                 2021.11.2
timm                     0.4.12
torch                    1.13.1
torchsummary             1.5.1
torchsummaryX            1.3.0
torchvision              0.14.1
torchviz                 0.0.2
tornado                  6.2
tqdm                     4.66.1
traitlets                5.9.0
ttach                    0.0.3
typing_extensions        4.5.0
urllib3                  1.26.16
wcwidth                  0.2.6
Werkzeug                 2.2.3
wheel                    0.38.4
wrapt                    1.15.0
zipp                     3.15.0

第一次在github上提问，期待作者解疑答惑
